src/Cli/dotnet/CommonOptions.cs

```diff
@@ -93,13 +98,53 @@ public static Argument DefaultToCurrentDirectory(this Argument arg)
 
         public static Option DebugOption() => new Option<bool>("--debug");
 
+        public static Option SelfContainedOption() =>
+            new ForwardedOption<bool>(
+                "--self-contained",
+                CommonLocalizableStrings.SelfContainedOptionDescription)
+            .ForwardAsSingle(o => $"-property:SelfContained={o}");
+
+        public static Option NoSelfContainedOption() =>
+            new ForwardedOption<bool>(
+                "--no-self-contained",
+                CommonLocalizableStrings.FrameworkDependentOptionDescription)
+            .ForwardAs("-property:SelfContained=false");
+
         public static bool VerbosityIsDetailedOrDiagnostic(this VerbosityOptions verbosity)
         {
             return verbosity.Equals(VerbosityOptions.diag) ||
                 verbosity.Equals(VerbosityOptions.diagnostic) ||
                 verbosity.Equals(VerbosityOptions.d) ||
                 verbosity.Equals(VerbosityOptions.detailed);
         }
+
+        public static void ValidateSelfContainedOptions(bool hasSelfContainedOption, bool hasNoSelfContainedOption, bool hasRuntimeOption, IEnumerable<string> projectArgs)
+        {
+            if (hasSelfContainedOption && hasNoSelfContainedOption)
+            {
+                throw new GracefulException(CommonLocalizableStrings.SelfContainAndNoSelfContainedConflict);
+            }
+
+            if (!(hasSelfContainedOption || hasNoSelfContainedOption) && hasRuntimeOption)
+            {
+                projectArgs = projectArgs.Any() ? projectArgs : new string[] { Directory.GetCurrentDirectory() };
+                foreach (var project in projectArgs)
+                {
+                    try
+                    {
+                        var msbuildProj = MsbuildProject.FromFileOrDirectory(new ProjectCollection(), project, false);
```

Comment:
Should we be evaluating the project in the CLI here?  What kind of a perf impact will that have?

@rainersigwald

Perhaps we should try to generate this warning from the project build itself.  That would mean it would be a NETSDK warning message with an error code, and could be ignored, which would also be nice.

Reply:
I added the error here because we wouldn't be able to detect when these properties were set on the command line vs by targets, but I suppose I could add an extra property with the CLI options to mark that it was set on the cli then move the error into targets. Do you want me to update with that approach?

Reply:
Evaluation is expensive enough that it seems like a good idea to push this into targets (but I don't have all the context on this particular thing fresh in my mind).

Reply:
Sounds like it would be a good idea to move it into the targets.  Though I think it's likely to generate the warning for each project built, as the properties we pass from the CLI are global properties which by default flow to referenced projects.

Reply:
Updated, moved to targets. @dsplaisted do you mind re-reviewing?